File: pidex/notifiers/telegram.py
import logging

import requests

from pidex.core.event import Event
from pidex.notifiers.base import BaseNotifier

logger = logging.getLogger(__name__)

TELEGRAM_API = "https://api.telegram.org/bot{token}/sendMessage"


class TelegramNotifier(BaseNotifier):
    def __init__(self, bot_token: str, chat_id: str):
        self._api_url = TELEGRAM_API.format(token=bot_token)
        self._chat_id = chat_id
        self._session = requests.Session()

    def send(self, event: Event) -> None:
        text = self._format_message(event)
        payload = {
            "chat_id": self._chat_id,
            "text": text,
            "parse_mode": "HTML",
            "disable_web_page_preview": True,
        }
        try:
            resp = self._session.post(self._api_url, json=payload, timeout=10)
            resp.raise_for_status()
            logger.info("Sent %s notification", event.event_type)
        except requests.RequestException:
            logger.exception("Failed to send Telegram message for %s", event.event_type)
            raise

    @staticmethod
    def _format_message(event: Event) -> str:
        severity_icon = {
            "INFO": "\u2139\ufe0f",
            "WARN": "\u26a0\ufe0f",
            "CRITICAL": "\U0001f6a8",
            "RECOVERED": "\u2705",
        }.get(event.severity, "")

        lines = [
            f"{severity_icon} <b>{event.title}</b>",
            f"<code>{event.message}</code>",
            f"\U0001f4c5 {event.timestamp.strftime('%Y-%m-%d %H:%M:%S')}",
            f"\U0001f3e0 {event.source}",
        ]
        return "\n".join(lines)

File: internal/notifier/telegram.go
package notifier

import (
	"bytes"
	"encoding/json"
	"fmt"
	"html"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/Shivamingale3/pi_dex/internal/core"
)

const maxRetries = 3

var severityIcon = map[string]string{
	core.SeverityInfo:     "\u2139\ufe0f",
	core.SeverityWarn:     "\u26a0\ufe0f",
	core.SeverityCritical: "\U0001f6a8",
	core.SeverityRecovered: "\u2705",
}

type TelegramNotifier struct {
	apiURL   string
	chatID   string
	client   *http.Client
	hostname string
}

func NewTelegramNotifier(token, chatID string) *TelegramNotifier {
	hostname, _ := os.Hostname()
	return &TelegramNotifier{
		apiURL:   fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", token),
		chatID:   chatID,
		client:   &http.Client{Timeout: 10 * time.Second},
		hostname: hostname,
	}
}

func (n *TelegramNotifier) Send(event core.Event) error {
	text := n.formatMessage(event)
	payload := map[string]any{
		"chat_id":                  n.chatID,
		"text":                     text,
		"parse_mode":               "HTML",
		"disable_web_page_preview": true,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	var lastErr error
	for attempt := range maxRetries {
		resp, err := n.client.Post(n.apiURL, "application/json", bytes.NewReader(body))
		if err != nil {
			lastErr = err
			log.Printf("Telegram send failed (attempt %d/%d): %v", attempt+1, maxRetries, err)
			time.Sleep(time.Duration(1<<attempt) * time.Second)
			continue
		}
		resp.Body.Close()
		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			log.Printf("Sent %s notification", event.EventType)
			return nil
		}
		lastErr = fmt.Errorf("HTTP %d", resp.StatusCode)
		log.Printf("Telegram send failed (attempt %d/%d): HTTP %d", attempt+1, maxRetries, resp.StatusCode)
		time.Sleep(time.Duration(1<<attempt) * time.Second)
	}

	return fmt.Errorf("telegram send failed after %d retries: %w", maxRetries, lastErr)
}

func (n *TelegramNotifier) formatMessage(event core.Event) string {
	icon := severityIcon[event.Severity]
	lines := []string{
		fmt.Sprintf("%s <b>%s | %s</b>", icon, event.Severity, html.EscapeString(event.Title)),
		html.EscapeString(event.Message),
		"",
		fmt.Sprintf("  Server    %s", n.hostname),
		fmt.Sprintf("  Source    %s", html.EscapeString(event.Source)),
		fmt.Sprintf("  Time      %s", event.Timestamp.Format("2006-01-02 15:04:05")),
		"",
		fmt.Sprintf("\u2014 PiDex v%s", core.Version),
	}
	var result string
	for i, line := range lines {
		if i > 0 {
			result += "\n"
		}
		result += line
	}
	return result
}
